runtime/compiler/trj9/arm/codegen/J9TreeEvaluator.cpp

```diff
@@ -0,0 +1,215 @@
+      TR::RegisterDependencyConditions *deps, TR::CodeGenerator *cg)
+   {
+   TR::Compilation * comp = cg->comp();
+   // non-heap objects cannot be marked
+   // Make sure we really should be here
+   TR::Options *options = comp->getOptions();
+   TR::Node *wrtbarNode = NULL;
+   bool definitelyHeapObject = false, definitelyNonHeapObject = false;
+   TR_WriteBarrierKind gcMode = comp->getOptions()->getGcMode();
+
+   if (node->getOpCodeValue() == TR::wrtbari || node->getOpCodeValue() == TR::wrtbar)
+      wrtbarNode = node;
+   else if (node->getOpCodeValue() == TR::ArrayStoreCHK)
+      wrtbarNode = node->getFirstChild();
+
+   if (wrtbarNode != NULL)
+      {
+      definitelyHeapObject = wrtbarNode->isHeapObjectWrtBar();
+      definitelyNonHeapObject = wrtbarNode->isNonHeapObjectWrtBar();
+      }
+
+   TR_ASSERT((gcMode == TR_WrtbarCardMark || gcMode == TR_WrtbarCardMarkAndOldCheck || gcMode == TR_WrtbarCardMarkIncremental) && !definitelyNonHeapObject,
+         "VMCardCheckEvaluator: Invalid call to cardCheckEvaluator\n");
+
+   if (!definitelyNonHeapObject)
+      {
+      TR::Register *metaReg = cg->getMethodMetaDataRegister();
+      TR::LabelSymbol *noChkLabel = generateLabelSymbol(cg);
+
+      // Balanced policy must always dirty the card table.
+      //
+      if (gcMode != TR_WrtbarCardMarkIncremental)
+         {
+         uint32_t base, rotate;
+
+         generateTrg1MemInstruction(cg, ARMOp_ldr, node, temp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(metaReg, offsetof(J9VMThread, privateFlags), cg));
+
+         // The value for J9_PRIVATE_FLAGS_CONCURRENT_MARK_ACTIVE is a generated value when VM code is created
+         // At the moment we are safe here, but it is better to be careful and avoid any unexpected behavoiur
+         // Make sure this falls within the scope of andis
+         //
+         TR_ASSERT(J9_PRIVATE_FLAGS_CONCURRENT_MARK_ACTIVE >= 0x00010000 && J9_PRIVATE_FLAGS_CONCURRENT_MARK_ACTIVE <= 0x80000000,
+               "Concurrent mark active Value assumption broken.");
+         constantIsImmed8r(J9_PRIVATE_FLAGS_CONCURRENT_MARK_ACTIVE, &base, &rotate);
+	 generateSrc1ImmInstruction(cg, ARMOp_tst, node, temp1Reg, base, rotate);
+         generateConditionalBranchInstruction(cg, node, ARMConditionCodeEQ, noChkLabel);
+         }
+
+      uintptr_t card_size_shift = trailingZeroes((uint32_t) options->getGcCardSize());
+
+      // temp3Reg = dstReg - heapBaseForBarrierRange0
+      generateTrg1MemInstruction(cg, ARMOp_ldr, node, temp3Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(metaReg, offsetof(J9VMThread, heapBaseForBarrierRange0), cg));
+      generateTrg1Src2Instruction(cg, ARMOp_sub, node, temp3Reg, dstReg, temp3Reg);
+
+      if (!definitelyHeapObject)
+         {
+         // if (temp3Reg >= heapSizeForBarrierRage0), object not in the heap
```

Comment:
Should that be "Range" rather than "Rage" ?  :-)

Reply:
This comes from the code for p.